stm32h747 camera support + stm32h7s related DT fixes

avolmat-st opened the pull request:
PR containing support of the STM32H747 camera (dcmi) support as well as some other fixes done.

In order to have this working, PR #95803 have to be merged as well.

The ST-B-CAMS-OMV shield is not yet functionnal with the STM32H747I-DISCO hence for the time being I am pushing this PR with the addition of the ST_STM32F4DIS_CAM shield support.
The OV5640 currently has a pixelclk configured too high, leading to the DCMI not able to capture everything and getting an partially broken image.

In addition to the STM32H747 bits, I also update few points regarding the STM32H7S and the ST-B-CAMS-OMV shield which I found while trying to enable the ST-B-CAMS-OMV with the platform.

I added the DCMI pinctrl within the stm32h747i-disco.dtsi file instead of the //m7 specific file. Indeed, I think this is where it belong since this is board related and not related to the point of using the M4 or the M7. Following this logic, I think that there are existing nodes containing pinctrl within the //m7 dts file which should also be moved into the dtsi file. This is a point to be discussed, maybe for another PR (or this one if this is preferred).

Commit:
boards: st: stm32h747i_disco: add camera support

Add connector, i2c and dcmi labels and pinctrl in order to allow
usage of a camera module shield with this the stm32h747i_disco board.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
boards: st: stm32h7b3i_dk: add dcmi pinctrl and remove duplicated i2c4

The pin alternate function settings are boards related so configure
the dcmi pinctrl within the board dts file in order to remove them
from the shield board specific overlay.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
shields: st_b_cams_omv_mb1683: remove i2c/dcmi/dmamux nodes from overlay

The i2c bus configuration and dcmi pins alternate function settings are
already done within the board dts since this is fixed for each board
so remove then from the shield overlay.
The dmamux node enabling is also removed since currently the DCMI/DMA
cannot rely on dmamux.

In the shield main overlay file, ensure that the i2c bus is properly
enabled.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
shields: st_stm32f4dis_cam: add stm32l4r9i_disco specific overlay

Move the selection of the dma channel and the sensor input clock into
a board specific overlay since this differ from a board to another.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
shields: st_stm32f4dis_cam: add a stm32h747i_disco//m7 overlay

Add an overlay allowing to use the stm32f4dis_cam shield on the
stm32h747i_disco board.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dts: st: l4: fix dcmi dma stream id

Channel ID should be numbered starting from 1.
Correct the dma properties of dcmi devices in L4 series
and use dma1 by default to make it same for all L4 series.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

avolmat-st commented:
Update the dma channel id for the L4 in the similar way as what has been done for the H7 on PR #95803

erwango commented on boards/shields/st_stm32f4dis_cam/boards/stm32l4r9i_disco.overlay:
```diff
@@ -0,0 +1,25 @@
+	status = "okay";
+};
+
+#define MCO1_SEL_SYSCLK 1
+
+#define MCO1_PRE_DIV_4 2
```
Not blocking: We should really get rid of these and provide definitions in `include/zephyr/dt-bindings/clock/stm32l4_clock.h` as done on other series.

erwango commented on boards/shields/st_stm32f4dis_cam/boards/stm32h747i_disco_stm32h747xx_m7.overlay:
```diff
@@ -0,0 +1,13 @@
+	status = "okay";
+};
+
+&dmamux1 {
+	status = "okay";
+};
```
You probably don't need dmamux anymore. Can you verify ?

erwango reviewed (commented):
Minor question, otherwise LGTM

Commit:
dma: stm32: dma_stm32_slot_to_channel is only applicable on STM32Fx

The dma_stm32_slot_to_channel function is only applicable on the
STM32Fx series since the LL_DMA_CHANNEL_x macros only exist for those
platforms among the one based on st,stm32-dma-v1.
While checking for !defined(CONFIG_DMAMUX_STM32) is ok to enable this
on STM32Fx, it prevents to disable dmamux on other platforms since
required symbols do not exist.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
dma: stm32: removal of unused stm32_dma_config_channel_function

Function stm32_dma_config_channel_function is unused hence remove it
from headers and dma_stm32_v1. In case of dma_stm32_v1, the related
configuration is already part of the LL_DMA_Init call done in
another place in dma_stm32.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
include: zephyr: dt-bindings: clock: l4: add MCO SEL/PRE macros

Add macros in order to select the MCO input clock and prescaler
value.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
include: zephyr: dt-bindings: clock: h7: add MCO SEL/PRE macros

Add macros in order to select the MCO input clock and prescaler
value.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

Commit:
shields: weact_ov2640_cam_module: use MCO macro from stm32 bindings

Switch to the usage of MCO configuration macros provided by
stm32h7_clock.h clock bindings.

Signed-off-by: Alain Volmat <alain.volmat@foss.st.com>

etienne-lms reviewed (commented):
Coding style comments only.

avolmat-st commented:
Thanks for the review @etienne-lms. Hopefully corrected all.

erwango commented on drivers/dma/dma_stm32_v1.c:
```diff
@@ -35,7 +35,7 @@ uint32_t dma_stm32_id_to_stream(uint32_t id)
 }
 
-#if !defined(CONFIG_DMAMUX_STM32)
+#if !defined(CONFIG_SOC_SERIES_STM32H7X) && !defined(CONFIG_SOC_SERIES_STM32MP1X)
```
Then, what about applying the same to `dma_stm32_slot_to_channel()` declaration in header:
https://github.com/zephyrproject-rtos/zephyr/blob/5fd0412d1db9018103d4814e8245948f2e0d4b43/drivers/dma/dma_stm32.h#L53-L55

[Out of scope of this PR]
This also raises the question on another user of this function: `stm32_dma_config_channel_function`, also defined under `#if !defined(CONFIG_DMAMUX_STM32)`, but this one is never called, so maybe it could be purely removed ?
@FRASTM can you have a look ?
